File: pdui_flutter/lib/widgets/pdui_text_form_field.dart
import 'package:flutter/material.dart';
import 'package:pdui_flutter/widgets/pdui_form.dart';

import '../proto_out/pdui-proto-out.pb.dart';
import 'pdui_widget.dart';

class PduiTextFormField extends PduiWidget {
  PBExpression pduiTextFormField;

  PduiTextFormField(this.pduiTextFormField);

  TextInputType _resolveTextInputType(PBKeyboardType pbKeyboardType) {
    switch (pbKeyboardType) {
      case PBKeyboardType.EMAIL_ADDRESS:
        return TextInputType.emailAddress;
      case PBKeyboardType.NUMBER:
        return TextInputType.number;
      case PBKeyboardType.DATE_TIME:
        return TextInputType.datetime;
      case PBKeyboardType.MULTILINE:
        return TextInputType.multiline;
      default:
        return TextInputType.text;
    }
  }

  @override
  Widget resolve() {
    String inputKey = pduiTextFormField.key;
    String formKey = pduiTextFormField.textFormField.formKey;
    bool obscureText = pduiTextFormField.textFormField.obscureText;

    var maxLength = pduiTextFormField.textFormField.maxLength > 0
        ? pduiTextFormField.textFormField.maxLength
        : null;

    var maxLines = !obscureText && pduiTextFormField.textFormField.maxLines > 0
        ? pduiTextFormField.textFormField.maxLines
        : 1;

    PduiFormController.formsInputValues[formKey]?[inputKey] = "";

    TextFormField textFormField = TextFormField(
      maxLength: maxLength,
      maxLines: maxLines,
      // minLength: pduiTextFormField.textFormField.minLength,
      obscureText: obscureText,
      keyboardType: _resolveTextInputType(
        pduiTextFormField.textFormField.keyboardType,
      ),
      decoration: InputDecoration(
        labelText: pduiTextFormField.textFormField.decoration.labelText,
      ),
      onChanged: (value) {
        PduiFormController.formsInputValues[formKey]?[inputKey] = value;
      },
      validator: (value) {
        if (value == null || value == "") {
          return "This field is required";
        } else {
          return null;
        }
      },
    );

    return textFormField;
  }
}

File: pdui-express/lib/widgets/PDUITextFormField.ts
import { PDUIWidget } from "../models/PDUIWidget";
import type { PBExpression } from "../proto-out/pdui-proto-out";
import { PDUIKeyboardType } from "./PDUIKeyboardType";
import type { PDUITextFormFieldDecoration } from "./PDUITextFormFieldDecoration";

export class PDUITextFormField extends PDUIWidget {
    formKey: string;
    keyboardType?: PDUIKeyboardType;
    decoration?: PDUITextFormFieldDecoration;
    obscureText?: boolean;
    notEmpty?: boolean;
    //TODO:
    // minLength?: number;
    maxLength?: number;
    maxLines?: number;
    //TODO:
    // validator: PDUIGenericFn;

    constructor({
        key,
        formKey,
        keyboardType,
        decoration,
        obscureText,
        notEmpty,
        // minLength,
        maxLength,
        maxLines,
    }: {
        key: string;
        formKey: string;
        keyboardType?: PDUIKeyboardType;
        decoration?: PDUITextFormFieldDecoration;
        obscureText?: boolean;
        notEmpty?: boolean;
        maxLength?: number;
        maxLines?: number;
    }) {
        //TODO: Validate things like: obscured fields can't be multiline...
        super({
            key,
            identity: "textFormField",
            payload: {
                oneofKind: "textFormField",
                textFormField: {
                    formKey: formKey,
                    keyboardType: keyboardType,
                    decoration: decoration ? decoration.toPB() : {},
                    obscureText: obscureText,
                    notEmpty: notEmpty,
                    maxLength: maxLength,
                    maxLines: maxLines,
                },
            },
        });

        this.formKey = formKey;
        this.keyboardType = keyboardType;
        this.decoration = decoration;
        this.obscureText = obscureText;
        this.notEmpty = notEmpty;
        this.maxLength = maxLength;
        this.maxLines = maxLines;
    }

    override toPB() {
        return {
            formKey: this.formKey,
            keyboardType: this.keyboardType,
            decoration: this.decoration ? this.decoration.toPB() : {},
            obscureText: this.obscureText,
            notEmpty: this.notEmpty,
            maxLength: this.maxLength,
            maxLines: this.maxLines,
        };
    }

    override toExpression(): PBExpression {
        return {
            key: this.key,
            identity: this.identity,
            payload: {
                oneofKind: "textFormField",
                textFormField: {
                    formKey: this.formKey,
                    keyboardType: this.keyboardType,
                    decoration: this.decoration ? this.decoration.toPB() : {},
                    obscureText: this.obscureText,
                    notEmpty: this.notEmpty,
                    maxLength: this.maxLength,
                    maxLines: this.maxLines,
                },
            },
        };
    }
}
